File: src/CannabisMath.Core/Potency/PotencyCalculator.cs
namespace CannabisMath.Core.Potency;

public static class PotencyCalculator
{
    private const decimal AcidToNeutralConversionFactor = 0.877m;

    public static decimal CalculateTotalThc(decimal thcPercent, decimal thcaPercent)
    {
        ValidateNonNegative(thcPercent, nameof(thcPercent));
        ValidateNonNegative(thcaPercent, nameof(thcaPercent));

        return thcPercent + (thcaPercent * AcidToNeutralConversionFactor);
    }

    public static decimal CalculateTotalCbd(decimal cbdPercent, decimal cbdaPercent)
    {
        ValidateNonNegative(cbdPercent, nameof(cbdPercent));
        ValidateNonNegative(cbdaPercent, nameof(cbdaPercent));

        return cbdPercent + (cbdaPercent * AcidToNeutralConversionFactor);
    }

    private static void ValidateNonNegative(decimal value, string parameterName)
    {
        if (value < 0)
        {
            throw new ArgumentOutOfRangeException(parameterName, "Value cannot be negative.");
        }
    }
}

File: js/src/potency/potencyCalculator.ts
export class PotencyCalculator {
  private static readonly CONVERSION_FACTOR = 0.877;

  static calculateTotalThc(thc: number, thca: number): number {
    this.validateNonNegative(thc, "thc");
    this.validateNonNegative(thca, "thca");

    return thc + thca * this.CONVERSION_FACTOR;
  }

  static calculateTotalCbd(cbd: number, cbda: number): number {
    this.validateNonNegative(cbd, "cbd");
    this.validateNonNegative(cbda, "cbda");

    return cbd + cbda * this.CONVERSION_FACTOR;
  }

  private static validateNonNegative(value: number, name: string) {
    if (value < 0) {
      throw new Error(`${name} cannot be negative`);
    }
  }
}